pkg/sql/contention/txnidcache/fifo_cache.go

```diff
@@ -0,0 +1,198 @@
+	blockPool *sync.Pool
+	capacity  contentionutils.CapacityLimiter
+
+	mu struct {
+		syncutil.RWMutex
+
+		data map[uuid.UUID]roachpb.TransactionFingerprintID
+
+		eviction evictionList
+	}
+}
+
+type evictionNode struct {
+	messageBlock
+	next *evictionNode
+}
+
+// evictionList is a singly-linked list of messageBlocks. The list is used to
+// implement FIFO eviction.
+type evictionList struct {
+	head *evictionNode
+	tail *evictionNode
+
+	// tailBlockIdx is an index pointing into the next empty slot in messageBlock
+	// stored in the tail pointer.
+	tailBlockIdx int
+}
+
+func newFIFOCache(pool *sync.Pool, capacity contentionutils.CapacityLimiter) *fifoCache {
+	c := &fifoCache{
+		blockPool: pool,
+		capacity:  capacity,
+	}
+
+	c.mu.data = make(map[uuid.UUID]roachpb.TransactionFingerprintID)
+	c.mu.eviction = evictionList{}
+	return c
+}
+
+// Add insert a messageBlock into the fifoCache.
+//
+// N.B. After Add() returns, the input messageBlock should not be reused. The
+// fifoCache sometimes directly hijack the input messageBlock to form the
+// eviction list and often even directly overwrite its content to minimize
+// memory allocation.
+func (c *fifoCache) Add(block *messageBlock) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	blockSize := 0
+	for i := range block {
+		if !block[i].valid() {
+			break
+		}
+
+		c.mu.data[block[i].TxnID] = block[i].TxnFingerprintID
+		blockSize++
+	}
+
+	c.mu.eviction.insertBlock(block, blockSize)
+
+	// Zeros out the block and put it back into the blockPool.
+	*block = messageBlock{}
+	c.blockPool.Put(block)
+
+	c.maybeEvictLocked()
+}
+
+func (c *fifoCache) Get(txnID uuid.UUID) (roachpb.TransactionFingerprintID, bool) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	fingerprintID, found := c.mu.data[txnID]
+	return fingerprintID, found
+}
+
+func (c *fifoCache) Size() int {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return len(c.mu.data)
+}
+
+func (c *fifoCache) maybeEvictLocked() {
+	for int64(len(c.mu.data)) > c.capacity() {
+		node := c.mu.eviction.removeFront()
+		if node == nil {
+			return
+		}
+
+		c.evictNodeLocked(node)
+
+		// Zero out the node and put it back into the pool.
+		*node = evictionNode{}
+		nodePool.Put(node)
+	}
+}
+
+// evictNodeLocked deletes all entries in the block from the internal map.
+func (c *fifoCache) evictNodeLocked(node *evictionNode) {
+	for i := 0; i < messageBlockSize; i++ {
+		if !node.messageBlock[i].valid() {
+			break
+		}
+
+		delete(c.mu.data, node.messageBlock[i].TxnID)
+	}
+}
+
+// insertBlock insert the messageBlock into the eviction list.
+func (e *evictionList) insertBlock(block *messageBlock, blockSize int) {
+	// Adding the new block when the list is empty.
+	if e.head == nil && e.tail == nil {
+		newNode := nodePool.Get().(*evictionNode)
+		copy(newNode.messageBlock[:], block[:blockSize])
+
+		e.head = newNode
+		e.tail = e.head
+		e.tailBlockIdx = blockSize
+
+		return
+	}
+
+	// Handle when the tail block is full.
+	if e.tail.isFull() {
```

Comment:
what if, instead of this method, you put a method on the `evictionNode` which was:

```go
func (e *evictionList) append(msgs []ResolveTxnID) {
    msgs = e.appendToTail(msgs)
    for len(msgs) > 0 {
        e.addNode()
        msgs = e.appendToTail(msgs)
    }
}

func (e *evictionList) addNode() {
    newNode := nodePool.Get().(*evictionNode)
    if e.head == nil {
        e.head = newNode
    } else {
        e.tail.next = newNode
    }
    e.tail = newNode
}

func (e *evictionList) appendToTail(msgs []ResolvedTxnID) (remaining []ResolvedTxnID) {
    if e.head == nil {
        return msgs
    }
    toCopy := messageBlockSize - e.tailBlockIdx
    if toCopy > len(msgs) {
       toCopy = len(msgs)
    }
    copy(e.tail.messageBlock[:], msgs[:toCopy])
    return msgs[toCopy:]
}
```